### ISSUE TITLE
io.py should be removed and/or handled differently

### ISSUE THREAD

**Issue by paskino** [vais-ral/CILViewer#162]
I guess the iDVC app is currently using the functionality in `utils/io.py` but we should remove this file.

**Comment by lauramurgatroyd** [comment]
Currently io.py takes an image file, reads it, possibly resampling or cropping, and returns a new image data, and a dictionary which contains info about the new image. There are also some options to write out files.

Currently...

Takes input:
-	Image file or list of image files
-	Empty vtkImageData – where the result will be stored
-	Whether to resample, and target size
-	Whether to crop on z axis, and new extent
-	Folder where to save the output image as a file
-	Option to write out numpy / raw
-	Dictionary to store info about the new image including vol_bit_depth and location of file
-	Method to run after conversion is complete

Then:
-	Identifies file type
-	If raw, opens raw input dialog, then waits for user to save settings
-	Creates a worker which runs the appropriate read method (usually a resample or cropped reader)
-	Updates a progress bar in a dialog on the interface whilst the worker runs
-	Updates input vtkImageData with the new image and dictionary with info about the image
-	Runs a method when worker is finished (method given in input).

Issues:
-	Relies on mainwindow having method to create progress window etc.
-	Shouldn’t need a mainwindow in order to run

**Instead I think we want to do the following:**

Create some class / method which takes an image file and args for settings.
-	It will identify which reader to use and return an image data
-	It will have the option to write out files, and user may specify the type
-	It will not rely on having a main window or start any Worker, instead another app can handle running it in a qprocess/worker
-	It will not launch a raw input dialog – it will rely on this info already being given in the input to the method – instead another app would need to launch the dialog in the case a raw image is chosen.
-  note we can still have the raw input dialog functionality in the viewer, just not in here!

This can be used by https://github.com/vais-ral/CILViewer/issues/165

https://github.com/vais-ral/CILViewer/issues/165 could take a config file or args, which would be read in and then these options would be passed as kwargs to our new class

**Comment by lauramurgatroyd** [comment]
We could have something like 

```python
class ImageReader(object):

    def __init__(self, **kwargs):
        '''
        Constructor
        
        Parameters
        ----------
        file_name: os.path or string, default None
            file name to read
        resample: bool, default False
            whether to resample
        crop: bool, default False
            whether to crop
        target_size: int
            target size after downsampling
        crop_extent:
            desired extent after cropping
                    
        '''
        # get kwargs then:
        self.set_up(file_name, resample, target_size, crop, crop_extent)

    def set_up(self,  file_name = None, resample=True, target_size=512**3, crop=False, crop_extent=None):

        self.file_name = file_name 
        # ... etc.
            
    def read(self):
        ''' reads self.file_name
            returns vtkImageData'''
        # identifies file type
        # uses appropriate reader based on file type and cropping or resampling

    def write(self, out_fname):
        ''' writes to a file the image data produced by the read method
        Parameters
        ----------
        outfname: str
            filename or path where to save output image file.
            File extension will determine the type of file written.
        '''

```

could do with a better name

**Comment by lauramurgatroyd** [comment]
When we are resampling the data, we want to still know the original image size.
So we need to return this information somewhere.

When we write out a downsampled image as a file we need to save this info somewhere as well, so maybe we need to write out some file to go along with it. This is easy in HDF5 because we can add some attributes, but with other types we would probably need a second file.

**Comment by lauramurgatroyd** [comment]
The resample readers modify both the origin and spacing of the vtkImageData so that the images are displayed correctly on the world.
This information needs to be retained when we write out an image file, so that when it is loaded in, this is placed correctly. There are some file formats where we can't add this in e.g. I think in numpy we can't save the origin, so again like above this could be saved in some additional file

**Comment by lauramurgatroyd** [comment]
Instead of having options for filetype to write out to, the write method will write out to a HDF5 file with:
- attributes of downsampled image including spacing, origin
- attributes of original image including size, possibly file location?

**Comment by paskino** [comment]
Yes, maybe this class will write to our own HDF5 file and optionally to TIFF stack + a metadata file (which may be a json file).

**Comment by lauramurgatroyd** [comment]
We need to decide on a convention for saving the information to the HDF5 file.

We'll be saving a single data set, with attributes about this, and also attributes of the original dataset this was generated from.

Should the data set be called: `'entry1/tomo_entry/data/data'` as in a nexus file?

We should have some basic attributes about the file similar to in CIL's [NexusDataWriter ](https://github.com/TomographicImaging/CIL/blob/376a0a85adaf400fb5276a2b6afc80983cf8b02a/Wrappers/Python/cil/io/NEXUSDataWriter.py#L117) and then attributes linked to the data. Something like:

```python
        with h5py.File(self.file_name, 'w') as f:
            
            # give the file some important attributes
            f.attrs['file_name'] = self.file_name
            f.attrs['cilviewer_version'] = version
            f.attrs['file_time'] = str(datetime.datetime.utcnow())
            f.attrs['creator'] = np.string_('<>.py')
            f.attrs['HDF5_Version'] = h5py.version.hdf5_version
            f.attrs['h5py_version'] = h5py.version.version
            
            # create the NXentry group
            nxentry = f.create_group('entry1/tomo_entry')
            nxentry.attrs['NX_class'] = 'NXentry'

            #create data entry
            ds_data = f.create_dataset('entry1/tomo_entry/data/data',shape, dtype)

            # data attributes:
            ds_data.attrs['original_fname'] = filepath to original data
            ds_data.attrs['original_shape'] = shape of  original  data
            ds_data.attrs['original_origin'] =
            ds_data.attrs['original_spacing'] =
            ds_data.attrs['resampled'] = bool
            ds_data.attrs['cropped'] = bool
            ds_data.attrs['spacing'] =
           ds_data. attrs['origin']=
    

```

**Comment by lauramurgatroyd** [comment]
We agreed that instead there should be a separate group for the original dilename, i.e. saved as entry1 and the other downsampled/cropped ones saved as entry2 etc.

**Comment by paskino** [comment]
I suppose this could be closed as the `io.py` has been removed?
